src/Linters/NoPHPArrayLiteralsLinter.hack

```diff
@@ -0,0 +1,155 @@
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST\Linters;
+
+use type Facebook\HHAST\{
+  ArrayCreationExpression,
+  ArrayIntrinsicExpression,
+  ArrayToken,
+  DictToken,
+  NodeList,
+  Node,
+  ElementInitializer,
+  IPHPArray,
+  LeftBracketToken,
+  LeftParenToken,
+  ListItem,
+  RightBracketToken,
+  RightParenToken,
+  Script,
+  VecToken,
+};
+use namespace HH\Lib\{C, Vec};
+use function Facebook\HHAST\Missing;
+
+final class NoPHPArrayLiteralsLinter extends AutoFixingASTLinter {
+  const type TNode = IPHPArray;
+  const type TContext = Script;
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    Script $_context,
+    IPHPArray $expr,
+  ): ?ASTLintError {
+    return new ASTLintError(
+      $this,
+      'Do not use PHP array literals, use Hack arrays instead.',
+      $expr,
+      () ==> $this->getFixedNode($expr),
+    );
+  }
+
+  public function getFixedNode(IPHPArray $expr): Node {
+    /*
+     * Cannot use getDescendantsOfType(ElementInitializer::class) because an
+     * associative array inside a list array would return a false positive.
+     */
+    $is_assoc = false;
+    foreach ($expr->getChildren() as $child) {
+      if ($child->isList()) {
+        foreach ($child->getChildren() as $item) {
+          if ($item is ListItem<_>) {
```

Comment:
for now, avoid `is` with generics in linters :( use `instanceof`

Performance of `is Foo<_>` currently heavily depends on using repo-authoritative mode